### PR TITLE
Add support for `Num.pow(Num)`

### PR DESCRIPTION
**Pull request by MarijnKoesen**
Fixes #.

Changes proposed in this pull request:
- Add support for `Num.pow(Num)`

- [ ] added an entry to the unreleased section of `CHANGES.md`

### CODE REVIEW

**Comment by edlins** [comment]
Note: after merge, refactor all/as-many-as-possible `Math.pow()` calls to `Num.pow()` calls.

**Comment by edlins** [comment]
I'm out today but this is the priority.

**Comment by team172011** [comment]
Merged because of high priority dependency

File: ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java [hunk 1]
```diff
@@ -181,6 +181,36 @@ public Num pow(int n) {
...
+     * @return <tt>this<sup>n</sup></tt>
+     * @see BigDecimal#pow(int, java.math.MathContext)
+     */
+    public Num pow(Num n) {
+        // Because BigDecimal.pow(BigDecimal) is not supported we need to find another way, we could do:
+        // Math.pow(a.doubleValue(), b.doubleValue()), but there is a better way:
+        // Perform X^(A+B)=X^A*X^B (B = remainder)
+        // As suggested: https://stackoverflow.com/a/3590314
+        BigDecimal result;
+        BigDecimal power = ((BigDecimalNum)n).delegate;
```
Comment (edlins) [inline ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:196]: Put this in `NumTest` and it fails on `ClassCastException`:
```
        Num fiveTest = numOf(5);
        Num fiveDouble = DoubleNum.valueOf(5);
        assertNumEquals(3125, fiveTest.pow(fiveDouble));
```

Comment (edlins) [reply ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:196]: Sometimes CCE is good.  Seems wrong here.  Open for discussion.  Any `Num` should be able to be raised to the `pow` of any other `Num` regardless of implementation.

File: ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java [hunk 2]
```diff
@@ -181,6 +181,36 @@ public Num pow(int n) {
...
+        int signOfPower = power.signum();
+        power = power.multiply(new BigDecimal(signOfPower));
```
Comment (edlins) [inline ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:198]: ~Why not power = power.abs()?~

Comment (edlins) [reply ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:198]: ~Or, however `power` is initially defined above, just add `.abs()` to the end of the definition.~
See my final comment on this file.

File: ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java [hunk 3]
```diff
@@ -181,6 +181,36 @@ public Num pow(int n) {
...
+
+        BigDecimal remainderOf2 = power.remainder(BigDecimal.ONE);
+        BigDecimal n2IntPart = power.subtract(remainderOf2);
```
Comment (edlins) [inline ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:201]: OOS, but there should be a `Num Num.floor()`...

File: ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java [hunk 4]
```diff
@@ -181,6 +181,36 @@ public Num pow(int n) {
...
+        if (signOfPower == -1) {
+            result = BigDecimal.ONE.divide(result, MATH_CONTEXT);
+        }
+
+        return new BigDecimalNum(result);
+    }
+
     /**
      * Returns the correctly rounded positive square root of the <code>double</code> value of this {@code Num}.
      * /!\ Warning! Uses the {@code StrictMath#sqrt(double)} method under the hood.
```
Comment (edlins) [inline ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:216]: I re-wrote this as:
```
        BigDecimal power = (((BigDecimalNum) n).delegate);
        BigDecimal remainderOf2 = power.remainder(BigDecimal.ONE);
        BigDecimal n2IntPart = power.subtract(remainderOf2);
        BigDecimal intPow = delegate.pow(n2IntPart.intValueExact(), MATH_CONTEXT);
        BigDecimal doublePow = new BigDecimal(Math.pow(delegate.doubleValue(), remainderOf2.doubleValue()));
        BigDecimal result = intPow.multiply(doublePow);
        return new BigDecimalNum(result);
```
And got the same unit test results.

File: ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java (hunk 1 above)
Comment (edlins) [reply ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:196]: Just double checked and this may actually be consistent with all of the other operations.  It just seems wrong to me that you can't
```
Num a = DoubleNum.valueOf("1.0");
Num b = BigDecimalNum.valueOf("2.0");
assertNumEquals("3.0", a.plus(b));
```
But I'm willing to let this go for now.

**Comment by edlins** [comment]
It would also be slick to see test cases for all of the overflow conditions.

File: ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java (hunk 1 above)
Comment (team172011) [reply ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:196]: > Any Num should be able to be raised to the pow of any other Num regardless of implementation.

I do not agree with this assumption. `Num#pow` could be overloaded with int/double, but different `Num` implementations should not be compatible in any way.

Comment (edlins) [reply ta4j-core/src/main/java/org/ta4j/core/num/BigDecimalNum.java:196]: Okay, I'll let it go.  :)  I now approve of line 196.

**Comment by edlins** [comment]
My current assessment is that the unnecessary operations on line 198 and 208 should be removed per my [comment](https://github.com/ta4j/ta4j/pull/212#discussion_r173636408) above.

**Comment by MarijnKoesen** [comment]
I think it would be nice to create a followup with the simplified implementation and tests 👍
